File: ruby/word-count/word_count.rb
class Phrase
  WORDS = /\b[\w']+\b/
  attr_reader :phrase
  def initialize(phrase)
    @phrase = phrase
  end

  def word_count
    phrase.scan(WORDS).each_with_object(Hash.new(0)) do |word, memo|
      memo[word.downcase] += 1
    end
  end
end

File: elixir/word-count/lib/word_count.ex
defmodule WordCount do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    words = split(sentence)

    words
    |> Enum.uniq()
    |> Enum.reduce(
      %{},
      fn word, counts ->
        Map.put_new(counts, word, occurrences(words, word))
      end
    )
  end

  def split(sentence) do
    sentence
    |> String.split(~r/[[:blank:][:punct:]]/u, trim: true)
    |> Enum.map(&String.downcase/1)
  end

  def occurrences(words, word) do
    words
    |> Enum.filter(&(&1 == word))
    |> Enum.count()
  end
end
